gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifyContentsPerExecutor.java

```diff
@@ -0,0 +1,269 @@
+      boolean isRefDroppedAfterCutoffTimeStamp =
+          droppedRefTime == null || droppedRefTime.compareTo(cutOffTimestamp) >= 0;
+      if (!isRefDroppedAfterCutoffTimeStamp) {
+        // reference is dropped before cutoff time.
+        // All the contents for all the keys are expired.
+        return new HashMap<>();
+      }
+      Predicate<CommitMeta> liveCommitPredicate =
+          commitMeta ->
+              // If the commit time is newer than (think: greater than or equal to) cutoff-time,
+              // then commit is live.
+              commitMeta.getCommitTime().compareTo(cutOffTimestamp) >= 0;
+
+      ImmutableGCStateParamsPerTask gcStateParamsPerTask =
+          ImmutableGCStateParamsPerTask.builder()
+              .api(api)
+              .reference(reference)
+              .liveCommitPredicate(liveCommitPredicate)
+              .bloomFilterSize(bloomFilterSize)
+              .build();
+
+      return walkLiveCommitsInReference(gcStateParamsPerTask);
+    }
+  }
+
+  private IdentifiedResult computeExpiredContents(
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap, Reference reference) {
+    try (NessieApiV1 api = GCUtil.getApi(gcParams.getNessieClientConfigs())) {
+      return walkAllCommitsInReference(api, reference, liveContentsBloomFilterMap);
+    }
+  }
+
+  private Map<String, ContentBloomFilter> walkLiveCommitsInReference(
+      GCStateParamsPerTask gcStateParamsPerTask) {
+    Map<String, ContentBloomFilter> bloomFilterMap = new HashMap<>();
+    Set<ContentKey> liveContentKeys = new HashSet<>();
+    try (Stream<LogResponse.LogEntry> commits =
+        StreamingUtil.getCommitLogStream(
+            gcStateParamsPerTask.getApi(),
+            Detached.REF_NAME,
+            gcStateParamsPerTask.getReference().getHash(),
+            null,
+            null,
+            OptionalInt.empty(),
+            true)) {
+      MutableBoolean foundAllLiveCommitHeadsBeforeCutoffTime = new MutableBoolean(false);
+      // commit handler for the spliterator
+      Consumer<LogResponse.LogEntry> commitHandler =
+          logEntry ->
+              handleLiveCommit(
+                  gcStateParamsPerTask,
+                  logEntry,
+                  bloomFilterMap,
+                  foundAllLiveCommitHeadsBeforeCutoffTime,
+                  liveContentKeys);
+      // traverse commits using the spliterator
+      GCUtil.traverseLiveCommits(foundAllLiveCommitHeadsBeforeCutoffTime, commits, commitHandler);
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+    return bloomFilterMap;
+  }
+
+  private IdentifiedResult walkAllCommitsInReference(
+      NessieApiV1 api,
+      Reference reference,
+      Map<String, ContentBloomFilter> liveContentsBloomFilterMap) {
+    IdentifiedResult result = new IdentifiedResult();
+    Instant commitProtectionTime =
+        (gcParams.getCommitProtectionDuration() == Duration.ZERO)
```

Comment:
From the javadoc of `Duration`:
_This is a value-based class; use of identity-sensitive operations (including reference equality (==), identity hash code, or synchronization) on instances of Duration may have unpredictable results and should be avoided. The equals method should be used for comparisons._

I know, that the _current implementation_ makes reference equality for `ZERO` possible, but that does not mean that it will be always the case.